cirq/google/optimizers/convert_to_sqrt_iswap.py

```diff
@@ -0,0 +1,283 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import cast, List, Optional
+
+import numpy as np
+
+import cirq
+import sympy
+
+SQRT_ISWAP = cirq.ISWAP**0.5
+SQRT_ISWAP_INV = cirq.ISWAP**-0.5
+
+
+# TODO: Combine this with the equivalent functions in google/gate_set.py
+# Or better yet, write a proper gate set so we don't need this in two places
+def _near_mod_n(e, t, n, atol=1e-8):
+    if isinstance(e, sympy.Basic):
+        return False
+    return abs((e - t + 1) % n - 1) <= atol
+
+
+def _near_mod_2pi(e, t, atol=1e-8):
+    return _near_mod_n(e, t, 2 * np.pi, atol=atol)
+
+
+class ConvertToSqrtIswapGates(cirq.PointOptimizer):
+    """Attempts to convert gates into ISWAP**-0.5 gates.
+
+    Since we have Z rotations and arbitrary XY rotations, we
+    can rely on cirq decomposition for one qubit gates and
+    need to only specify special decompositions for two qubit gates.
+
+    Currently natively specified gates are CZPowGate, ISwapPowGate,
+    and FSimGate.  This will also support gates that decompose into
+    the above gates.
+    """
+
+    def __init__(self, ignore_failures=False) -> None:
+        """
+        Args:
+            ignore_failures: If set, gates that fail to convert are forwarded
+                unchanged. If not set, conversion failures raise a TypeError.
+        """
+        super().__init__()
+        self.ignore_failures = ignore_failures
+
+    def _convert_one(self, op: cirq.Operation) -> cirq.OP_TREE:
+        """
+        Decomposer intercept:  Let cirq decompose one-qubit gates,
+        intercept on 2-qubit gates if they are known gates.
+        """
+        if isinstance(op, cirq.GlobalPhaseOperation):
+            return []
+
+        gate = op.gate
+
+        if len(op.qubits) != 2:
+            return NotImplemented
+
+        if isinstance(gate, cirq.CZPowGate):
+            if isinstance(gate.exponent, sympy.Basic):
+                return cphase_symbols_to_sqrt_iswap(op.qubits[0], op.qubits[1],
+                                                    gate.exponent)
+            else:
+                return cphase_to_sqrt_iswap(op.qubits[0], op.qubits[1],
+                                            gate.exponent)
+        if isinstance(gate, cirq.SwapPowGate):
+            return swap_to_sqrt_iswap(op.qubits[0], op.qubits[1], gate.exponent)
+        if isinstance(gate, cirq.ISwapPowGate):
+            return iswap_to_sqrt_iswap(op.qubits[0], op.qubits[1],
+                                       gate.exponent)
+        if isinstance(gate, cirq.FSimGate):
+            return fsim_gate(op.qubits[0], op.qubits[1], gate.theta, gate.phi)
+
+        return NotImplemented
+
+    def convert(self, op: cirq.Operation) -> List[cirq.Operation]:
+
+        def on_stuck_raise(bad):
```

Comment:
I'm not sure the performance implication of defining a local function like this, but it might be faster to make this an internal method on the class instead so it doesn't get redefined on each call. Might be worth benchmarking to see if that makes any difference when optimizing typical circuits.

Reply:
Not sure that it matters, but we don't really need an inline function anyway, so I just moved it out.